### PR TITLE
check for unixODBC in `dbConnect(odbc(), ...)`

### LINKED ISSUE #782: consider moving `odbc_config` check for unixODBC to `dbConnect<OdbcDriver>`
#753 added a check inside of `databricks()` to ensure that `odbc_config` can find unixODBC on macOS:

https://github.com/r-dbi/odbc/blob/903702c229ad64fcb8b3ad6346f3d12fc77484f5/R/utils.R#L179-L188

We [had wondered](https://github.com/r-dbi/odbc/pull/753/#discussion_r1496317243) if that check should just go inside of `dbConnect()` but decided not to as it seemed relatively invasive/ad-hoc. We might consider having this check kick in for all drivers on macOS/Linux in a later release of the package.

### PR DESCRIPTION
**Pull request by simonpcouch**
This PR adds a call to `locate_install_unixodbc()` to the main connection path, now that we've "piloted" it in the macOS x Databricks connection path for a release without issue. Closes #782.

### CODE REVIEW

File: R/utils.R [hunk 1]
```diff
@@ -278,27 +271,68 @@ configure_spark <- function(call = caller_env()) {
...
+    "/lib",
     "/usr/local/lib",
     "/opt/homebrew/lib",
     "/opt/homebrew/opt/unixodbc/lib"
   )
 
   list.files(
     common_dirs,
-    pattern = "libodbcinst\\.dylib$",
+    pattern = "libodbcinst\\.(dylib|so)$",
```
Comment (hadley) [inline R/utils.R:294]: Use `libodbcinst_extension()` here?

File: R/utils.R [hunk 2]
```diff
@@ -278,27 +271,68 @@ configure_spark <- function(call = caller_env()) {
...
+      unixodbc_prefix <- system(x, intern = TRUE, ignore.stderr = TRUE)
+      return(paste0(unixodbc_prefix, "/libodbcinst.", libodbcinst_extension()))
+    },
+    error = function(e) {},
+    warning = function(w) {}
+  )
+  character()
+}
+
+libodbcinst_extension <- function() {
```
Comment (hadley) [inline R/utils.R:311]: Might be a little nicer to make this `libodbcinst_filename`? (e.g return `libodbcinst.dylib`)

File: R/utils.R [hunk 3]
```diff
@@ -278,27 +271,68 @@ configure_spark <- function(call = caller_env()) {
...
+  if (is_macos()) {
```
Comment (hadley) [inline R/utils.R:312]: Here, where the if branches are "even", I'd do:

```R
if (x) {
  a
} else {
  b
}```

File: R/utils.R [hunk 4]
```diff
@@ -278,27 +271,68 @@ configure_spark <- function(call = caller_env()) {
...
+  )
+}
+
+prompt_install_unixodbc <- function() {
+  if (is_macos()) {
+    return("Please install unixODBC using Homebrew with `brew install unixodbc`.")
+  }
+
+  "Please install unixODBC and retry."
+}
```
Comment (hadley) [inline R/utils.R:334]:
```suggestion
error_install_unixodbc <- function(call) {
  abort(
    c(
      "Unable to locate the unixODBC driver manager.",
      i = if (is_macos()) "Please install unixODBC using Homebrew with `brew install unixodbc`.",
      i = if (!is_macos()) "Please install unixODBC and retry."
    ),
    call = call
  )
}
```

Maybe?

**Review by hadley — APPROVED** [review]
Love how we're continuing to grind down these annoying failure modes 😄

**Comment by simonpcouch** [comment]
Much cleaner all around, thanks!

**Pull request merged**